File: tests/st/ops/cpu/test_tensoradd.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import numpy as np
import pytest

import mindspore.common.dtype as mstype
import mindspore.nn as nn
from mindspore import Tensor, context
from mindspore.ops import operations as P

context.set_context(mode=context.GRAPH_MODE, device_target='CPU')

class TensorAdd(nn.Cell):
    def __init__(self):
        super(TensorAdd, self).__init__()
        self.add = P.Add()

    def construct(self, x, y):
        res = self.add(x, y)
        return res


@pytest.mark.level0
@pytest.mark.platform_x86_cpu
@pytest.mark.env_onecard
def test_tensor_add():
    x0 = Tensor(np.random.uniform(-2, 2, (2, 3, 4, 4)).astype(np.float32))
    y0 = Tensor(np.random.uniform(-2, 2, (1, 1, 1, 1)).astype(np.float32))
    x1 = Tensor(np.random.uniform(-2, 2, (1, 3, 1, 4)).astype(np.float32))
    y1 = Tensor(np.random.uniform(-2, 2, (2, 3, 4, 4)).astype(np.float32))
    x2 = Tensor(np.random.uniform(-2, 2, (2, 3, 4, 4)).astype(np.float32))
    y2 = Tensor(2, mstype.float32)
    x3 = Tensor(2, mstype.float32)
    y3 = Tensor(2, mstype.float32)
    x4 = Tensor(np.random.uniform(-2, 2, (4)).astype(np.float32))
    y4 = Tensor(np.random.uniform(-2, 2, (4, 4)).astype(np.float32))
    add = TensorAdd()
    out = add(x0, y0).asnumpy()
    exp = x0.asnumpy() + y0.asnumpy()
    diff = np.abs(out - exp)
    err = np.ones(shape=exp.shape) * 1.0e-5
    assert np.all(diff < err)
    assert out.shape == exp.shape

    out = add(x1, y1).asnumpy()
    exp = x1.asnumpy() + y1.asnumpy()
    diff = np.abs(out - exp)
    err = np.ones(shape=exp.shape) * 1.0e-5
    assert np.all(diff < err)
    assert out.shape == exp.shape

    out = add(x2, y2).asnumpy()
    exp = x2.asnumpy() + y2.asnumpy()
    diff = np.abs(out - exp)
    err = np.ones(shape=exp.shape) * 1.0e-5
    assert np.all(diff < err)
    assert out.shape == exp.shape

    out = add(x3, y3).asnumpy()
    exp = x3.asnumpy() + y3.asnumpy()
    diff = np.abs(out - exp)
    err = np.ones(shape=exp.shape) * 1.0e-5
    assert np.all(diff < err)
    assert out.shape == exp.shape

    out = add(x4, y4).asnumpy()
    exp = x4.asnumpy() + y4.asnumpy()
    diff = np.abs(out - exp)
    err = np.ones(shape=exp.shape) * 1.0e-5
    assert np.all(diff < err)
    assert out.shape == exp.shape

File: tests/cxx_st/model/test_tensor_add.cc
/**
 * Copyright 2020 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <string>
#include <vector>
#include "common/common_test.h"
#include "include/api/model.h"
#include "include/api/serialization.h"
#include "include/api/context.h"

using namespace mindspore;

static const char tensor_add_file[] = "/home/workspace/mindspore_dataset/tensor_add/tensor_add.mindir";
static const std::vector<float> input_data_1 = {1, 2, 3, 4};
static const std::vector<float> input_data_2 = {2, 3, 4, 5};
static const std::vector<float> input_data_shape5_1 = {1, 2, 3, 4, 5};
static const std::vector<float> input_data_shape5_2 = {2, 3, 4, 5, 6};

class TestTensorAdd : public ST::Common {
 public:
  TestTensorAdd() {}
};

TEST_F(TestTensorAdd, BuildOptionOutputType_Only310) {
  GlobalContext::SetGlobalDeviceTarget(kDeviceTypeAscend310);

  auto ascend310_context = std::make_shared<ModelContext>();
  ModelContext::SetOutputType(ascend310_context, DataType::kNumberTypeUInt8);
  auto graph = Serialization::LoadModel(tensor_add_file, ModelType::kMindIR);
  Model tensor_add((GraphCell(graph)), ascend310_context);
  Status ret = tensor_add.Build();
  ASSERT_TRUE(ret == kSuccess);

  // prepare input
  std::vector<MSTensor> inputs;
  std::vector<MSTensor> outputs;
  std::vector<MSTensor> origin_inputs = tensor_add.GetInputs();
  inputs.emplace_back(origin_inputs[0].Name(), origin_inputs[0].DataType(), origin_inputs[0].Shape(),
                      input_data_1.data(), sizeof(float) * input_data_1.size());
  inputs.emplace_back(origin_inputs[1].Name(), origin_inputs[1].DataType(), origin_inputs[1].Shape(),
                      input_data_2.data(), sizeof(float) * input_data_2.size());

  // infer
  ret = tensor_add.Predict(inputs, &outputs);
  ASSERT_TRUE(ret == kSuccess);

  // print
  for (auto &o : outputs) {
    ASSERT_EQ(o.DataType(), DataType::kNumberTypeUInt8);
    const uint8_t *p = reinterpret_cast<const uint8_t *>(o.MutableData());
    for (size_t i = 0; i < o.DataSize() / sizeof(uint8_t); ++i) {
      ASSERT_LE(std::abs(static_cast<float>(p[i]) - (input_data_1[i] + input_data_2[i])), 1e-4);
    }
  }
}
